### PR TITLE
Slack remodel

### PR DESCRIPTION
**Pull request by abelsiqueira**
Built on top of #115, please review only the last commit.

This adds `SlackModel(::ADNLPModel)` which returns `ADNLPModel`.

### CODE REVIEW

File: src/slack_model.jl [hunk 1]
```diff
@@ -68,16 +68,23 @@ function slack_meta(meta :: NLPModelMeta)
...
-function SlackModel(model :: AbstractNLPModel)
+"""
+    ConvertToSlackModel(model)
+
+Convert an NLPModel or NLSModel to a SlackModel, without using specialized information
+about the type of `model`.
+
+You may want `SlackModel(model)` instead.
+"""
+function ConvertToSlackModel(model :: AbstractNLPModel)
```
Comment (dpo) [inline src/slack_model.jl]: Why would we ever want this?

File: src/slack_model.jl [hunk 2]
```diff
@@ -372,3 +392,26 @@ function hess_op_residual!(nls :: SlackModel, x :: AbstractVector, i :: Int, Hiv
...
+  c(x) = begin
+    cx = nlp.c(@view x[1:n])
+    cx[nlp.meta.jlow] .-= @view x[n+1:n+nlow]
+    cx[nlp.meta.jupp] .-= @view x[n+nlow+1:n+nlow+nupp]
+    cx[nlp.meta.jrng] .-= @view x[n+nlow+nupp+1:n+nlow+nupp+nrng]
+    return cx
+  end
+
+  return ADNLPModel(meta, deepcopy(nlp.counters), f, c)
+end
```
Comment (dpo) [inline src/slack_model.jl]: Yes, that's the idea! I think this is the way to go.

File: src/slack_model.jl (hunk 1 above)
Comment (abelsiqueira) [reply src/slack_model.jl]: I'm not sure we will but currently, it's used for testing on `test/test_slack_model.jl`. Possibly won't need it after this is finished.

File: src/feasibility_residual.jl [hunk 3]
```diff
@@ -5,11 +5,16 @@ export FeasibilityResidual
...
-    s.t c(x) = 0
+    s.t cℓ ≤ c(x) ≤ cu
+        bℓ ≤   x  ≤ bu
 
-by defining the function F(x) = c(x). If the problem has
-bounds on the variables or more constraints, an error
-is thrown.
+by creating slack variables s and defining F(x,s) = c(x) - s. The resulting NLS problem is
+
+    min ¹/₂‖c(x) - s‖²
```
Comment (dpo) [inline src/feasibility_residual.jl]: It seems f(x) disappeared?!

Comment (abelsiqueira) [reply src/feasibility_residual.jl]: Yes, `f` is ignored on for the NLS model. I could change to `min 0`.

**Comment by abelsiqueira** [comment]
Dropping in favor of #115

**Pull request closed without merging**